### PR TITLE
Add Opt Out field and MailChimp integration

### PR DESCRIPTION
**Pull request by JeremySkinner**
These commits begin the work on adding Opt Out features for communication, as well as mailchimp integration for emails.

- Applies patch to drupal core for fixing an issue with hook entity_field_storage_info not working in D8
- Adds Communications tab to contacts dashboard
- Adds Commnucation profile type/bundle
- Creates/configures the crm_opt_out field 
- Creates a new glue module (contacts_mailchimp):
  - Uses the Mailchimp module (a patch is needed to handle email field on related entities - see https://www.drupal.org/project/mailchimp/issues/2949125)
  - Automatically unsubscribes any mailchimp lists when the crm_opt_out field is set
  - Automatically does the reverse (sets the crm_opt_out field to true when unsubscribe webhook is received from mailchimp)

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -38,7 +38,8 @@
     "patches": {
       "drupal/core": {
         "#2865059: Multiple machine name elements from same source": "https://www.drupal.org/files/issues/2865059-13-allow_multiple_machine_name_per_source.patch",
-        "#736066: ajax.js insert command sometimes wraps content in a div": "https://www.drupal.org/files/issues/736066-296.patch"
+        "#736066: ajax.js insert command sometimes wraps content in a div": "https://www.drupal.org/files/issues/736066-296.patch",
+        "#2584729: Hook entity_field_storage_info is broken": "https://www.drupal.org/files/issues/core-entity-sql-content-entity-storage-2583111-1_0.patch"
       },
       "drupal/ctools": {
         "#2667652: Option to expose filters in block on views block display": "https://www.drupal.org/files/issues/ctools-option_to_expose-2667652-3.patch",
```
Comment (yanniboi) [inline composer.json:45]: Was there not a mailchimp patch as well?

File: config/install/contacts.contact_tab.crm_communications_tab.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+uuid: 5800747c-04df-4555-902f-b233214033a7
+langcode: en
+status: true
```
Comment (yanniboi) [inline config/install/contacts.contact_tab.crm_communications_tab.yml]: We don't tend to include uuids for config that goes in contributed modules. @andrewbelcher is that still accurate?

File: config/install/contacts.contact_tab.crm_communications_tab.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  profile_crm_communications:
+    id: 'typed_data_entity_relationship:entity:user:profile_crm_communications'
+    name: profile_crm_communications
+    source: user
+blocks:
+  contacts_entity_profile_crm_communications:
+    id: 'contacts_entity:profile-crm_communications'
+    label: 'Communications Profile block'
+    provider: contacts
+    label_display: '0'
```
Comment (yanniboi) [inline config/install/contacts.contact_tab.crm_communications_tab.yml]: Lets change the label to 'Communication preferences'.

File: modules/contacts_mailchimp/contacts_mailchimp.module [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    if (isset($entity->original)) {
+      $original = $entity->original->get($field_name)->value;
+    }
+    else {
+      $original = NULL;
+    }
+
+    // TODO: Investigate making this asynchronous.
+    if (($original === 0 || $original == NULL) && $opt_out === 1 && $entity->uid->entity->hasField('mail')) {
+      $email = $entity->uid->entity->mail->value;
```
Comment (yanniboi) [inline modules/contacts_mailchimp/contacts_mailchimp.module:30]: Not many ways of doing asynchronous in Drupal. We could consider moving it to a cron queue for performance, but its probably fine for now.

**Review by yanniboi — CHANGES_REQUESTED** [review]
Looking good so far. A couple of tweaks.

File: composer.json (hunk 1 above)
Comment (JeremySkinner) [reply composer.json:45]: @yanniboi Yes - this is only needed if you install contacts_mailchimp though, so should it go in the composer.json for the contacts module, or would there need to be a new composer.json specifically for contacts_mailchimp?

File: config/install/contacts.contact_tab.crm_communications_tab.yml (hunk 2 above)
Comment (JeremySkinner) [reply config/install/contacts.contact_tab.crm_communications_tab.yml]: Removed the uuid.

File: config/install/contacts.contact_tab.crm_communications_tab.yml (hunk 3 above)
Comment (JeremySkinner) [reply config/install/contacts.contact_tab.crm_communications_tab.yml]: Applied renaming

File: config/install/contacts.contact_tab.crm_communications_tab.yml (hunk 2 above)
Comment (andrewbelcher) [reply config/install/contacts.contact_tab.crm_communications_tab.yml]: I think the current consensus is that UUIDs *should* be included in default configuration so you can see if different config entities are being managed (e.g. 2 modules provide the same profile type, UUID can tell you which one you have).
